### PR TITLE
Add hub config for geospatial workshop

### PR DESCRIPTION
**Pull request by sgibson91**
This PR adds a config file to deploy a hub for the Geospatial Workshop being held in Ghana later this month.

Done:

- Deploys a daskhub
- Uses Google authentication
- list of admins
- user limits for dask
- URL setup in NameCheap

Needs:

- Review and merge

related to: https://github.com/2i2c-org/pilot-hubs/issues/473

### CODE REVIEW

File: config/hubs/pangeo-hubs.cluster.yaml
Comment (sgibson91) [reply config/hubs/pangeo-hubs.cluster.yaml:26]: Because I just copy-pasted the config from 2i2c-staging and I wasn't sure if it would be needed for the geospatial workshop, but also didn't want to lose the structure in case it _was_ required

**Comment by sgibson91** [comment]
I have pushed a commit with folks gmail account for admin access (mostly copy-pasted from configs on the 2i2c cluster). If your email is not there and you'd like it to be, please feel free to add it!

**Comment by sgibson91** [comment]
I have tried to set some limits for singleuser and dask gateway by borrowing config from pangeo:

- singleuser: https://github.com/pangeo-data/pangeo-cloud-federation/blob/d4f868e4d5c1ea92675facdabff86d49f31c7253/pangeo-deploy/values.yaml#L19-L25
- dask gateway: https://github.com/pangeo-data/pangeo-cloud-federation/blob/e0b8afdd5571a8a1e68b4b7a33b789703a97ac7f/deployments/icesat2/config/common.yaml#L149-L154

Would love to know if I'm on the right lines with this in terms of (i) is this correct config? (ii) will it achieve Paige's desires in https://github.com/2i2c-org/pilot-hubs/issues/473#issuecomment-879492933?

File: config/hubs/pangeo-hubs.cluster.yaml
Comment (yuvipanda) [reply config/hubs/pangeo-hubs.cluster.yaml:26]: Let's get rid of this for now, we can always copy paste it back if needed.

File: config/hubs/pangeo-hubs.cluster.yaml
Comment (yuvipanda) [inline config/hubs/pangeo-hubs.cluster.yaml:2]: This will probably not be pullable for this repo. Let's explicitly specify an image from github.com/pangeo-data/pangeo-docker-images/ under `singleuser` instead?

File: config/hubs/pangeo-hubs.cluster.yaml
Comment (yuvipanda) [inline config/hubs/pangeo-hubs.cluster.yaml:1]: Since this is on the *old* pangeo account, let's just use the name of the project as the name of the cluster? Otherwise we'll end up with a new pangeo cluster on the new account, and it'll get confusing. This might mean changes for the terraform workspace too, though.

Comment (sgibson91) [reply config/hubs/pangeo-hubs.cluster.yaml:1]: I deployed this cluster into a terraform workspace called `pangeo-test` rather than `pangeo-hubs`. I'll rename these files to `pangeo-181919`

**Comment by sgibson91** [comment]
In order to set the URL/DNS stuff, I tried deploying the hub with `proxy.https.enabled = false` using the deployer locally, but ran into the following error

```
auth0.v3.exceptions.Auth0Error: 403: You reached the limit of entities of this type for this tenant.
```

https://auth0.com/docs/policies/entity-limit-policy?_gl=1*1ulg0hl*rollup_ga*ODY5Njk2MDcyLjE2MjM4NTg3NDc.*rollup_ga_F1G3E656YZ*MTYyNjM1NjE4Ny4zLjEuMTYyNjM1NjE5OS40OA

This is a bug in creating duplicate auth0 apps

**Comment by sgibson91** [comment]
I'm having trouble with the NFS server for deployment.

- I created a VM like the `nfs-server-01` VM under `two-eye-two-see`
- On that VM, I created `/etc/exports` and wrote `/export/home-01 10.0.0.0/8(all_squash,anonuid=1000,anongid=1000,no_subtree_check,rw,sync)` to it

However, the `nfs-share-creator` is still in `ContainerCreating` mode with `FailedMount` events for the volume `"home-base"`. I'm not sure what to do next 😕

**Comment by damianavila** [comment]
> Would love to know if I'm on the right lines with this in terms of (i) is this correct config? (ii) will it achieve Paige's desires in #473 (comment)?

I think you are going in the right direction, but I am not a Dask guy :wink:
@yuvipanda, do you know if the options that are available here: https://github.com/2i2c-org/pilot-hubs/blob/master/hub-templates/daskhub/values.yaml#L183 are somehow "enforced/restricted" by the limits @sgibson91 wants to use in the config file?

> I'm having trouble with the NFS server for deployment.

IIRC, @yuvipanda had some issues in the past related to an "insecure" flag

> NFS server was again set up manually, and needed the insecure
flag - even though other hubs are setup the same way and didn't
need this. NFS situation needs to be sorted.

Not sure if that is relevant here... but I just remember it 😛  !

**Comment by sgibson91** [comment]
I think a lot of this is that I'm totally new to setting up NFS servers! I've just managed to work out that it helps to have the NFS packages installed on the VM! 😝

**Comment by sgibson91** [comment]
I have been playing around and the hub is definitely not happy. The hub is not reachable by it's external IP or the URL http://coessing.pangeo.2i2c.cloud

**Comment by damianavila** [comment]
> The hub is not reachable by it's external IP or the URL http://coessing.pangeo.2i2c.cloud

I just tried that URL and I was able to, at least, reach it...

**Comment by sgibson91** [comment]
Thanks @damianavila! I should be more patient with DNS things 😆 I will come back to this today

**Comment by sgibson91** [comment]
Ha, this was 💯 me not being patient enough with DNS stuff! Hub check now passes and I can log in! 🎉

File: config/hubs/pangeo-181919.cluster.yaml
Comment (damianavila) [inline config/hubs/pangeo-181919.cluster.yaml:29]: Why we are using LoadBalancer here?
I saw that in AWS-based clusters (and it makes sense, AFAIK) but the only other GCP cluster where this is used is on meom cluster... wondering why that is not needed in other clusters such as the 2i2c or cloudbank ones?

**Comment by sgibson91** [comment]
> Quick general question, we usually have staging and prod hubs on each cluster we create, ie. https://github.com/2i2c-org/pilot-hubs/blob/master/config/hubs/meom-ige.cluster.yaml
> I am only seeing one hub in this config, is that intended?

Yeah. I guess because I'm envisioning both this hub being brought down after the workshop, _and_ the cluster it's on being destroyed once the appropriate constraints have been lifted/amended on the new Pangeo account.

File: config/hubs/pangeo-181919.cluster.yaml
Comment (sgibson91) [reply config/hubs/pangeo-181919.cluster.yaml:29]: This is because this is a daskhub which sets a Cluster IP so Dask Gateway is accessed through JupyterHub: https://github.com/2i2c-org/pilot-hubs/blob/ec14dd8f13de016000cf679ad134648d20814cfb/hub-templates/daskhub/values.yaml#L200-L201

Without _also_ setting LoadBalancer, I was not getting an external IP from the proxy-public service - it was all just the cluster IP.

Comment (damianavila) [reply config/hubs/pangeo-181919.cluster.yaml:29]: OK, but that section you referenced belongs to a traefik config on the dask-gateway: https://github.com/2i2c-org/pilot-hubs/blob/master/hub-templates/daskhub/values.yaml#L97-L201 and not the proxy: https://github.com/2i2c-org/pilot-hubs/blob/master/hub-templates/basehub/values.yaml#L76, right? (or am I most likely missing something? :wink: )

> Without setting LoadBalancer, I was not getting an external IP from the proxy-public service - it was all just the cluster IP.

Ok, that explains why you set up the LoadBalancer in the proxy section, I still do not get why that is not needed in other clusters containing daskhubs.

File: secrets/pangeo-181919.json
Comment (choldgraf) [inline secrets/pangeo-181919.json:5]: is it a problem that I can see the text for all of these? Or are they properly encoded? :-)

**Review by choldgraf — COMMENTED** [review]
This looks good to me! I'll defer to @damianavila to make sure that his questions are resolved!

File: secrets/pangeo-181919.json
Comment (sgibson91) [reply secrets/pangeo-181919.json:5]: They are encoded as prefaced by the `ENC` part of the value. Compare, for example, with the key for the 2i2c cluster: https://github.com/2i2c-org/pilot-hubs/blob/master/secrets/2i2c.json I think what makes `sops` a little more clever than, say, `git-crypt` is that it encrypts the actual values, not just the whole file. So you can visually check the _structure_ is as expected without also needing to see the secret part.

File: config/hubs/pangeo-181919.cluster.yaml
Comment (sgibson91) [reply config/hubs/pangeo-181919.cluster.yaml:29]: @yuvipanda and I ended up cribbing from the carbon plan cluster which does have this config https://github.com/2i2c-org/pilot-hubs/blob/e38f58c9c11e16928308bb9f9c401171e60fd3e1/config/hubs/carbonplan.cluster.yaml#L115-L117

Comment (damianavila) [reply config/hubs/pangeo-181919.cluster.yaml:29]: That's ok, my general question is about why this is needed on this pangeo (and meom) cluster but it is not needed in other GCP clusters. In any case, my question should not be a blocker here, and I will create a ticket to follow up with that discussion.

**Review by damianavila — APPROVED** [review]
I have one remaining question but I will open a new issue to follow up on that discussion. 
AFAIK, this one is already deployed and working.
LGTM.

File: config/hubs/pangeo-181919.cluster.yaml
Comment (sgibson91) [inline config/hubs/pangeo-181919.cluster.yaml:61]:
```suggestion
              name: quay.io/2i2c/coessing-image
              tag: TBA
```

Waiting for https://github.com/2i2c-org/coessing-image/pull/3 to be merged and the image to be built before adding the tag and merging this one

Comment (sgibson91) [reply config/hubs/pangeo-181919.cluster.yaml:61]: **Update:** Added ref `5a8a765b7ff9` to config for bespoke image

File: config/hubs/pangeo-181919.cluster.yaml
Comment (sgibson91) [inline config/hubs/pangeo-181919.cluster.yaml:61]:
```suggestion
              tag: "9894ec5ce239"
```

**Comment by sgibson91** [comment]
Ok, this is all up-to-date now, I'm gonna merge!

**Comment by choldgraf** [comment]
🚀🚀🚀

**Pull request merged**
